### PR TITLE
Calendar - Meeting Query

### PR DESCRIPTION
**Pull request by AniMachaidze**
Optimized algorithm for finding available times for required attendees and added function for finding times that also works for most number of optional attendees possible. Also, added test cases to check correctness.

### CODE REVIEW

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 1]
```diff
@@ -11,134 +11,200 @@
...
+      if (busyOptionals.size() > 0) {
+        if (busyTimesForOptionals.containsKey(eventTimeRange)) {
+          busyTimesForOptionals.get(eventTimeRange).addAll(busyOptionals);
+        } else {
+          busyTimesForOptionals.put(eventTimeRange, busyOptionals);
         }
+      }
+    }
+
+    List<Set<String>> setsOfIptionals = new ArrayList<>();
```
Comment (oydang) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:95]: Nit: spelling, `setsOfOptionals`. Also, what this variable represents isn't very clear.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 2]
```diff
@@ -11,134 +11,200 @@
...
+
+    return freeTimes;
   }
 
-  /** Finds avaliable times for the given atteendee list, input collection and event duration */
-  private Collection<TimeRange> getAvailableTimes(
-      Collection<String> attendees, Collection<Event> events, int duration) {
-    // TODO: Make this method more optimal
-    Collection<TimeRange> availableTimes = new ArrayList<>();
+  /** Calculates for each minute in a day how many optional attendee are free */
```
Comment (oydang) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:73]: So this should return a list `j` of size 1440, with `j[i]` meaning the number of optional attendees are free at minute `i`? The function comment is a little unclear.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 3]
```diff
@@ -11,134 +11,200 @@
...
-        if (end == TimeRange.END_OF_DAY) {
-          availableTimes.add(
-              TimeRange.fromStartEnd(availablePeriodStart, availablePeriodEnd, true));
-        }
-      } else {
+    for (int i = 0; i <= TimeRange.END_OF_DAY; i++) {
+      numberOfFreeOptionals.add(totalOptionals - setsOfIptionals.get(i).size());
+    }
+
+    numberOfFreeOptionals.add(-1);
```
Comment (oydang) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:114]: This is to indicate the end of the ArrayList? I don't think this is necessary.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 4]
```diff
@@ -11,134 +11,200 @@
...
+      overlaps.add(new ArrayList<TimeRange>());
+    }
+
+    // for each available time for required attendees find chuncks of times with same number of free
+    // optionals and populates overlaps
+    for (TimeRange availableTime : availableTimes) {
+      int start = availableTime.start();
+      int end = availableTime.end();
+
+      int number = numberOfOptionals.get(start);
```
Comment (oydang) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:142]: Make this variable name more descriptive.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 5]
```diff
@@ -11,134 +11,200 @@
...
-        if (availablePeriodEnd != -1) {
-          availableTimes.add(
-              TimeRange.fromStartEnd(availablePeriodStart, availablePeriodEnd, false));
+    return numberOfFreeOptionals;
+  }
+
+  /**
+   * Calculates for each number of optional attendees, chich free time ranges are overlaping with
+   * available times for required attendees and returns overlaping free times with higest number
+   * possible
```
Comment (astrelni) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:122]: Nit: spelling: chich => which, higest => highest. Add a period at the end.

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java [hunk 6]
```diff
@@ -487,9 +495,244 @@ public void noTimeWithOnlyOptionals() {
...
+
+    Collection<TimeRange> actual = query.query(events, request);
+    Collection<TimeRange> expected =
+        Arrays.asList(
+            TimeRange.fromStartEnd(TimeRange.START_OF_DAY, TIME_0830AM, false),
+            TimeRange.fromStartEnd(TIME_1000AM, TIME_1100AM, false),
+            TimeRange.fromStartEnd(TIME_1330AM, TimeRange.END_OF_DAY, true));
+
+    Assert.assertEquals(expected, actual);
+  }
```
Comment (oydang) [inline walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:737]: I noticed none of your tests use an event that has multiple attendees. It'd be good to modify at least one of the events in at least one of the tests to have more than 1 attendee.

**Pull request merged**